lib/Monads/trace/Trace_More_RG.thy

```diff
@@ -0,0 +1,685 @@
+\<comment> \<open>FIXME MC: delete since we won't use K for rg? If we don't delete, it needs to assume the G
+lemma K_valid[wp]:
+  "\<lbrace>K P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>_. K P\<rbrace>"
+  by (simp add: valid_def)\<close>
+
+lemma rg_exI_tuple:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>(rv,rv') s0 s. Q x rv rv' s0 s\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>(rv,rv') s0 s. \<exists>x. Q x rv rv' s0 s\<rbrace>"
+  by (fastforce simp: validI_def)
+
+lemma rg_ex_all:
+  "(\<forall>x. \<lbrace>P x\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>) = \<lbrace>\<lambda>s0 s. \<exists>x. P x s0 s\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  apply (rule iffI)
+   apply (fastforce simp: validI_def)+
+  done
+
+lemma rg_imp_eq_substR:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace> \<Longrightarrow> \<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. rv = x \<longrightarrow> Q x s0 s\<rbrace>,\<lbrace>E\<rbrace>"
+  by (fastforce simp add: validI_def validIE_def split: sum.splits)
+
+lemma rg_split_bind_case_sum:
+  assumes x: "\<And>rv. \<lbrace>E rv\<rbrace>,\<lbrace>R\<rbrace> g rv \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+             "\<And>rv. \<lbrace>S rv\<rbrace>,\<lbrace>R\<rbrace> h rv \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  assumes y: "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace>,\<lbrace>E\<rbrace>"
+  shows      "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>= case_sum g h \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>"
+  apply (rule rg_seq_ext[OF _ y[unfolded validIE_def]])
+  apply (wpsimp wp: x split: sum.splits)
+  done
+
+lemma rg_split_bind_case_sumE:
+  assumes x: "\<And>rv. \<lbrace>S' rv\<rbrace>,\<lbrace>R\<rbrace> g rv \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+             "\<And>rv. \<lbrace>S rv\<rbrace>,\<lbrace>R\<rbrace> h rv \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  assumes y: "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f \<lbrace>G\<rbrace>,\<lbrace>S\<rbrace>,\<lbrace>S'\<rbrace>"
+  shows      "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> f >>= case_sum g h \<lbrace>G\<rbrace>,\<lbrace>Q\<rbrace>,\<lbrace>E\<rbrace>"
+  apply (unfold validIE_def)
+  apply (rule rg_seq_ext[OF _ y[unfolded validIE_def]])
+  apply (wpsimp wp: x[unfolded validIE_def] split: sum.splits)
+  done
+
+lemma assertE_tsp:
+  "\<lbrace>P\<rbrace>,\<lbrace>R\<rbrace> assertE Q \<lbrace>G\<rbrace>,\<lbrace>\<lambda>rv s0 s. Q \<and> P s0 s\<rbrace>,\<lbrace>E\<rbrace>"
+  by (wpsimp wp: assertE_wp)
+
+\<comment> \<open>FIXME MC: delete? The only reason to keep is so that the RG version can get simplified, otherwise
+           wp will lift it to a hoare triple\<close>
+lemma gets_inv_rg[simp]:
```

Comment:
I haven't really tested it out, but if we removed this lemma then I think the behaviour would be that `wp` would lift it to a hoare triple, and then `simp` would resolve it with `gets_inv`, the hoare equivalent to this lemma.

The only benefit I can think of for additionally having this version is so that anything adapted from existing proofs that just used `simp` would keep on working, without needing to add that `wp` step.

Reply:
I'd remove it -- we're going to have to adjust proofs anyway and that should make things a bit more uniform